src/lib/fountain/concurrent_fountain_decoder_sink.h

```diff
@@ -35,14 +35,37 @@ class concurrent_fountain_decoder_sink
 		return _decoder.num_done();
 	}
 
+	std::vector<std::string> get_done() const
+	{
+		std::lock_guard<std::mutex> lock(_readMutex);
+		return _done;
+	}
+
+	std::vector<double> get_progress() const
+	{
+		std::lock_guard<std::mutex> lock(_readMutex);
+		return _progress;
+	}
+
+	void update_status()
+	{
+		// called under the writeMutex -- don't hold this for long!
+		std::lock_guard<std::mutex> lock(_readMutex);
+		_done = _decoder.get_done();
+		_progress = _decoder.get_progress();
+	}
+
 	void process()
 	{
-		if (_mutex.try_lock())
+		if (_writeMutex.try_lock())
 		{
+			// maybe a 2nd mutex for retrieving _decoder.has_file()??
 			std::string buff;
 			while (_backlog.try_dequeue(buff))
 				_decoder << buff;
```

Comment:
With a more flexible write interface, we would probably be able to return progress/completeness up out of the write. However, the (rough) adherence to the iostreams api prevents this. (`write()` presents the same problem)

Returning the values from the write is attractive, because it means we only need 1 mutex instead of 2 here. But I'm not going to rewrite all the streams -- yet?